discovery/hetzner/robot.go

```diff
@@ -0,0 +1,120 @@
+// Copyright 2020 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package hetzner
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	config_util "github.com/prometheus/common/config"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/discovery/refresh"
+	"github.com/prometheus/prometheus/discovery/targetgroup"
+)
+
+const (
+	hetznerRobotLabelPrefix    = hetznerLabelPrefix + "robot_"
+	hetznerLabelRobotProduct   = hetznerRobotLabelPrefix + "product"
+	hetznerLabelRobotCancelled = hetznerRobotLabelPrefix + "cancelled"
+)
+
+// Discovery periodically performs Hetzner Robot requests. It implements
+// the Discoverer interface.
+type robotDiscovery struct {
+	*refresh.Discovery
+	client   *http.Client
+	port     int
+	endpoint string
+}
+
+// newRobotDiscovery returns a new robotDiscovery which periodically refreshes its targets.
+func newRobotDiscovery(conf *SDConfig, logger log.Logger) (*robotDiscovery, error) {
+	d := &robotDiscovery{
+		port:     conf.Port,
+		endpoint: conf.robotEndpoint,
+	}
+
+	rt, err := config_util.NewRoundTripperFromConfig(conf.HTTPClientConfig, "hetzner_sd", false, false)
+	if err != nil {
+		return nil, err
+	}
+	d.client = &http.Client{
+		Transport: rt,
+		Timeout:   time.Duration(conf.RefreshInterval),
+	}
+
+	return d, nil
+}
+func (d *robotDiscovery) refresh(ctx context.Context) ([]*targetgroup.Group, error) {
+	resp, err := d.client.Get(d.endpoint + "/server")
+	if err != nil {
+		defer resp.Body.Close()
```

Comment:
the defer should happend when wee is not nil also please call io.Copy(ioutil.Discard, resp.Body)